vrx_gz/src/WildlifeScoringPlugin.cc

```diff
@@ -0,0 +1,810 @@
+
+  // Parse a new buoy.
+  while (buoyElem)
+  {
+    // The buoy's model.
+    if (!buoyElem->HasElement("name"))
+    {
+      gzerr << "Unable to find <buoys><buoy><name> element in SDF."
+            << std::endl;
+      return false;
+    }
+
+    const std::string buoyModelName = buoyElem->Get<std::string>("name");
+
+    // The buoy's goal.
+    if (!buoyElem->HasElement("goal"))
+    {
+      gzerr << "Unable to find <buoys><buoy><goal> element in SDF."
+            << std::endl;
+      return false;
+    }
+
+    const std::string buoyGoal = buoyElem->Get<std::string>("goal");
+    if (!this->AddBuoy(buoyModelName, buoyGoal, _ecm))
+      return false;
+
+    // Parse the next buoy.
+    buoyElem = buoyElem->GetNextElement("buoy");
+  }
+
+  return true;
+}
+
+//////////////////////////////////////////////////
+bool WildlifeScoringPlugin::Implementation::AddBuoy(const std::string &_name,
+  const std::string &_goal, sim::EntityComponentManager &_ecm)
+{
+  auto entity = sim::entitiesFromScopedName(_name, _ecm);
+  if (entity.empty())
+  {
+    gzerr << "Unable to find entity [" << _name << "]" << std::endl;
+    return false;
+  }
+
+  BuoyGoal buoyGoal;
+  if (_goal == "avoid")
+    buoyGoal = BuoyGoal::AVOID;
+  else if (_goal == "circumnavigate_clockwise")
+    buoyGoal = BuoyGoal::CIRCUMNAVIGATE_CLOCKWISE;
+  else if (_goal == "circumnavigate_counterclockwise")
+    buoyGoal = BuoyGoal::CIRCUMNAVIGATE_COUNTERCLOCKWISE;
+  else
+  {
+    gzerr << "Unknown <goal> value: [" << _goal << "]" << std::endl;
+    return false;
+  }
+
+  // Save the new buoy.
+  this->buoys.push_back(
+    Buoy(_name, *entity.begin(), buoyGoal, this->engagementDistance, _ecm));
+
+  return true;
+}
+
+//////////////////////////////////////////////////
+void WildlifeScoringPlugin::Implementation::PublishAnimalLocations(
+  const sim::UpdateInfo &_info, sim::EntityComponentManager &_ecm)
+{
+  msgs::Pose_V geoPathMsg;
+
+  auto stamp = sim::convert<msgs::Time>(_info.simTime);
+  geoPathMsg.mutable_header()->mutable_stamp()->CopyFrom(stamp);
+
+  for (auto const &buoy : this->buoys)
+  {
+    // Conversion from Gazebo Cartesian coordinates to spherical.
+    auto comp = _ecm.Component<sim::components::WorldPose>(buoy.entity);
+    if (!comp)
+    {
+      gzerr << "Unable to find WorldPose component for entity ["
+            << buoy.entity << "]" << std::endl;
+      return;
+    }
+
+    math::Pose3d pose = comp->Data();
+
+    auto in = math::SphericalCoordinates::CoordinateType::GLOBAL;
+    auto out = math::SphericalCoordinates::CoordinateType::SPHERICAL;
+    auto latlon = this->sc.PositionTransform(pose.Pos(), in, out);
+    latlon.X(GZ_RTOD(latlon.X()));
+    latlon.Y(GZ_RTOD(latlon.Y()));
+
+    const math::Quaternion<double> orientation = pose.Rot();
+
+    // Fill the message.
+    auto geoPoseMsg = geoPathMsg.add_pose();
+
+    // header->stamp
+    geoPoseMsg->mutable_header()->mutable_stamp()->CopyFrom(stamp);
```

Comment:
Do we need to do this twice? See line 706.

Reply:
It's a bit redundant but some embedded fields also have header so I filled it for consistency.